src/parser.rs

```diff
@@ -725,19 +725,87 @@ impl<'a> Parser<'a> {
     }
 
     fn parse_list(&mut self) -> Result<ast::Expression, ParseError> {
+        let location = self.location;
         self.consume_token("[")?;
         self.consume_whitespace();
-        let elements = self.parse_sequence(']')?;
-        self.consume_token("]")?;
-        Ok(ast::ListComprehension { elements }.into())
+        if let Ok(_) = self.consume_token("]") {
+            return Ok(ast::ListLiteral { elements: vec![] }.into());
+        }
+        let first_element = self.parse_expression()?;
+        self.consume_whitespace();
+        if let Ok(_) = self.consume_token("]") {
+            let elements = vec![first_element];
+            Ok(ast::ListLiteral { elements }.into())
+        } else if let Ok(_) = self.consume_token(",") {
+            self.consume_whitespace();
+            let other_elements = self.parse_sequence(']')?;
+            self.consume_whitespace();
+            self.consume_token("]")?;
+            let elements = vec![first_element]
+                .into_iter()
+                .chain(other_elements)
+                .collect();
```

Comment:
I think this is simpler as:

```suggestion
            let mut elements = self.parse_sequence(']')?;
            self.consume_whitespace();
            self.consume_token("]")?;
            elements.insert(0, first_element);
```